test/apps/iodemo/io_demo.cc

```diff
@@ -0,0 +1,464 @@
+
+        listen((const struct sockaddr*)&listen_addr, sizeof(listen_addr));
+        for (;;) {
+            try {
+                progress();
+            } catch (const std::exception &e) {
+                std::cerr << e.what();
+            }
+        }
+    }
+
+    void handle_io_read_request(UcxConnection* conn, const iomsg_hdr_t *hdr) {
+        // send data
+        conn->send_data(buffer(), opts().data_size, hdr->sn);
+
+        // send response as data
+        IoMessage *response = new IoMessage(opts().iomsg_size, IO_COMP, hdr->sn,
+                                            0);
+        conn->send_data(response->buffer(), opts().iomsg_size, hdr->sn,
+                        response);
+    }
+
+    void handle_io_write_request(UcxConnection* conn, const iomsg_hdr_t *hdr) {
+        conn->recv_data(buffer(), opts().data_size, hdr->sn,
+                        new IoWriteResponseCallback(this, conn, hdr->sn));
+    }
+
+    virtual void dispatch_io_message(UcxConnection* conn, const void *buffer) {
+        const iomsg_hdr_t *hdr = reinterpret_cast<const iomsg_hdr_t*>(buffer);
+
+        if (opts().verbose) {
+            std::cout << "got io message " << io_op_names[hdr->op]
+                      << " sn "   << hdr->sn << " data size " << hdr->data_size
+                      << " conn " << conn    << std::endl;
+        }
+
+        if (hdr->op == IO_READ) {
+            handle_io_read_request(conn, hdr);
+        } else if (hdr->op == IO_WRITE) {
+            handle_io_write_request(conn, hdr);
+        }
+    }
+};
+
+
+class DemoClient : private P2pDemoCommon {
+public:
+    class IoReadResponseCallback : public UcxCallback {
+    public:
+        IoReadResponseCallback(long *counter, size_t iomsg_size) :
+            _counter(0), _io_counter(counter), _buffer(malloc(iomsg_size)) {
+        }
+
+        ~IoReadResponseCallback() {
+            free(_buffer);
+        }
+
+        virtual void operator()() {
+            /* wait data and response completion */
+            if (++_counter < 2) {
+                return;
+            }
+
+            ++(*_io_counter);
+            delete this;
+        }
+
+        void* buffer() {
+            return _buffer;
+        }
+
+    private:
+        long  _counter;
+        long* _io_counter;
+        void* _buffer;
+    };
+
+    DemoClient(const options_t& test_opts) :
+        P2pDemoCommon(test_opts),
+        _num_sent(0), _num_completed(0) {
+    }
+
+    void do_io_read(UcxConnection *conn, uint32_t sn) {
+        ++_num_sent;
+        send_io_message(conn, IO_READ, sn);
+        IoReadResponseCallback *response =
+                new IoReadResponseCallback(&_num_completed, opts().iomsg_size);
+        conn->recv_data(buffer(), opts().data_size, sn, response);
+        // receive response as data to ensure ordering, and count completions
+        conn->recv_data(response->buffer(), opts().iomsg_size, sn, response);
```

Comment:
maybe change io_read response to use io_message instead of data? since it's based on counter anyway now..

Reply:
do you mean remove `IoReadResponseCallback::_buffer` and use extended `_data_buffer` also for response?

Reply:
talke offline - no need